torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -591,9 +575,78 @@ def _update_trainable(self) -> None:
             self._sync_param_groups(self.optim.param_groups, self.param_groups)
 
         if self.parameters_as_bucket_view:
             self._setup_flat_buffers()
+
+    def _verify_params(self, params: Any) -> None:
+        r"""
+        Verifies the type of ``params`` and initializes ``self._all_params``
+        if ``params`` is valid.
+
+        While :class:`optim.Optimizer <torch.optim.Optimizer>` allows
+        ``params`` to be an iterable of :class:`dict` s, currently
+        ``ZeroRedundancyOptimizer`` strictly requires ``params`` to be an
+        iterable of :class:`torch.Tensor` s.
+
+        Raises:
+            TypeError: ``params`` has an invalid type.
+            ValueError: ``params`` is empty.
+        """
+        if (isinstance(params, torch.Tensor)):
+            raise TypeError("params argument should be an iterable of "
+                            "Tensors, but got " + torch.typename(params))
+        try:
+            self._all_params = list(params)
+        except TypeError:
+            raise TypeError("params argument should be an iterable of "
+                            "Tensors, but got " + torch.typename(params))
+        if (len(self._all_params) == 0):
+            raise ValueError("ZeroRedundancyOptimizer got an empty parameter "
+                             "list")
+        for param in self._all_params:
+            if (not isinstance(param, torch.Tensor)):
+                raise TypeError("params argument should be an iterable of"
+                                "Tensors, but got an iterable containing " +
+                                torch.typename(param))
+
+    def _verify_spsd(self) -> None:
+        r"""
+        Verifies that ZeRO is being used under the single-process single-
+        device regime where a process operates exclusively on a full model
+        replica on a single device.
+
+        The function assumes that ``self._all_params`` has been initialized.
+
+        Raises:
+            ValueError: ``params`` contains parameters across multiple
+                devices.
+        """
+        assert len(self._all_params) > 0
+        device = self._all_params[0].device
+        for param in self._all_params[1:]:
+            if (param.device != device):
+                raise ValueError("ZeroRedundancyOptimizer assumes single-"
+                                 "process single-device (SPSD); make sure that"
+                                 "all parameters are on the same device")
+
+    def _verify_same_dtype(self) -> None:
+        r"""
+        Verifies that all parameters are of the same dtype.
+
+        The function assumes that ``self._all_params`` has been initialized.
+
+        Raises:
+            ValueError: ``params`` contains parameters of varying dtypes.
+
+        NOTE: This function can be removed once support for varying dtype is
+        added.
+        """
+        assert len(self._all_params) > 0
```

Comment:
Do we need to check this again if its already checked during init?

Reply:
Good point. This is not needed here since it is redundant with the previous check. 

I included it mostly out of past habit. It partially captures the precondition of the function and arguably future-proofs the function against changes to the `__init__()` implementation. 

Either way, I removed the `assert` and instead added in the comment that `self._all_params` is assumed to be non-empty.